### PR TITLE
[Mempool] Support network compression and maximum network batch bytes.

### PR DESCRIPTION
**Pull request by JoshLind**
### Description

This PR offers a number of small changes (mainly to mempool):
1. Clean up some of the network traits (to avoid applications defining methods they don't use).
2. Add network compression to mempool. In forge we're seeing 3x-4x traffic reduction. I suspect this to be similar in other environments (e.g., devnet) as this is what I observed with state sync 😄 There seems to be no performance regression when compared to recent PRs.
3. Updated mempool to take into account the max message sizes for the networking stack. Here, we limit the number of transactions sent per batch to other peers.
4. Cleaned up some small logic around `BatchId` serialize. We were previously serializing this field twice, which makes no sense to me 🤷 
5. Added two unit tests: (i) one to test the network byte limit chunking; and (ii) another to test that batches are correctly being limited by the max batch size in the config. As far as I can tell, this test is missing.

### Test Plan
New and existing tests.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3134)

### CODE REVIEW

File: types/src/transaction/mod.rs [hunk 1]
```diff
@@ -450,9 +460,10 @@ impl fmt::Debug for SignedTransaction {
             "SignedTransaction {{ \n \
              {{ raw_txn: {:#?}, \n \
              authenticator: {:#?}, \n \
+             bytes: {:#?}, \n \
              }} \n \
              }}",
-            self.raw_txn, self.authenticator
+            self.raw_txn, self.authenticator, self.bytes
```
Comment (gregnazario) [inline types/src/transaction/mod.rs:466]: I wouldn't add this, it'll add a lot of noise

File: types/src/account_config/resources/crsn.rs [hunk 2]
```diff
@@ -13,7 +13,7 @@ use serde::{Deserialize, Serialize};
 
 use crate::event::EventHandle;
 
-#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
 pub enum AccountSequenceInfo {
     Sequential(u64),
     CRSN { min_nonce: u64, size: u64 },
```
Comment (gregnazario) [inline types/src/account_config/resources/crsn.rs:19]: Lol I thought we killed CRSN?

File: mempool/src/shared_mempool/network.rs [hunk 3]
```diff
@@ -126,17 +126,11 @@ impl ApplicationNetworkSender<MempoolSyncMsg> for MempoolNetworkSender {
...
+        _timeout: Duration,
     ) -> Result<MempoolSyncMsg, RpcError> {
-        fail_point!("mempool::send_to", |_| {
-            Err(anyhow::anyhow!("Injected error in mempool::send_rpc").into())
-        });
-        let protocol = ProtocolId::MempoolRpc;
-        self.inner
-            .send_rpc(recipient, protocol, req_msg, timeout)
-            .await
+        unimplemented!("Shared mempool only supports direct send messages!");
```
Comment (gregnazario) [inline mempool/src/shared_mempool/network.rs:133]: That's so sad :(. Never forget

File: mempool/src/core_mempool/transaction_store.rs [hunk 4]
```diff
@@ -428,27 +430,38 @@ impl TransactionStore {
...
             if let Some(txn) = self
                 .transactions
                 .get(&address)
                 .and_then(|txns| txns.get(&sequence_number))
             {
-                batch.push(txn.txn.clone());
-                if let TimelineState::Ready(timeline_id) = txn.timeline_state {
-                    last_timeline_id = timeline_id;
+                let transaction_bytes = txn.txn.raw_txn_bytes_len() as u64;
+                if (batch_total_bytes + transaction_bytes) > self.max_batch_bytes {
```
Comment (gregnazario) [inline mempool/src/core_mempool/transaction_store.rs:453]: saturating add?  Though I guess max batch bytes should never get close to that

File: mempool/src/core_mempool/transaction_store.rs [hunk 5]
```diff
@@ -428,27 +430,38 @@ impl TransactionStore {
...
+                    break; // The batch is full
+                } else {
+                    batch.push(txn.txn.clone());
+                    batch_total_bytes += transaction_bytes;
```
Comment (gregnazario) [inline mempool/src/core_mempool/transaction_store.rs:457]: similar thoughts

**Review by gregnazario — APPROVED** [review]

File: types/src/account_config/resources/crsn.rs (hunk 2 above)
Comment (zekun000) [reply types/src/account_config/resources/crsn.rs:19]: it's killed on move layer, but I guess we never update the mempool

File: types/src/transaction/mod.rs (hunk 1 above)
Comment (zekun000) [reply types/src/transaction/mod.rs:466]: yeah, don't print out raw bytes, we can print out the len

Comment (JoshLind) [reply types/src/transaction/mod.rs:466]: Sounds good! Updated the code 😄

File: mempool/src/core_mempool/transaction_store.rs (hunk 4 above)
Comment (JoshLind) [reply mempool/src/core_mempool/transaction_store.rs:453]: Aah, thanks!

**Pull request merged**